### PR TITLE
multi: store bool to determine retransmission ordering

### LINKED ISSUE #4613: Node forced closed a channel with zapread.com, not sure why
### Background
My node force-closed a channel with zapread.com, it was a good and active channel, I am not sure why my node decided to force-close the channel, some relevant things that I found, full log attached
```
2020-09-02 04:58:55.181 [ERR] HSWC: ChannelLink(563360:2528:1): failing link: ChannelPoint(25194f104e65a27cfd6867ac8e8708c9cb898e1212a51ff1862ffb03cc1e340b:1): received error from peer: chan_id=0b341ecc03fb2f86f11fa512128e89cbc908878eac6768fd7ca2654e104f1924, err=internal error with error: remote error
2020-09-02 04:58:55.185 [ERR] HSWC: ChannelLink(563360:2528:1): link failed, exiting htlcManager
2020-09-02 04:58:55.185 [INF] HSWC: ChannelLink(563360:2528:1): exited
```
.
.
.
```
2020-09-02 12:45:07.278 [INF] DISC: GossipSyncer(037659a0ac8eb3b8d0a720114efc861d3a940382dcfa1403746b4f8f6b2e8810ba): applying gossipFilter(start=0001-01-01 00:00:00 +0000 UTC, end=0001-01-01 00:00:00 +0000 UTC)
2020-09-02 12:45:07.279 [INF] DISC: GossipSyncer(0231eccc6510eb2e1c97c8a190d6ea096784aa7c358355442055aac8b20654f932): applying gossipFilter(start=2020-09-02 12:45:07.279317998 +0000 UTC m=+408745.563033846, end=2156-10-09 19:13:22.279317998 +0000 UTC m=+4295376040.563033846)
2020-09-02 12:45:37.279 [INF] DISC: Broadcasting 265 new announcements in 18 sub batches
2020-09-02 12:46:03.069 [INF] CRTR: Processed channels=0 updates=232 nodes=0 in last 59.998613183s
2020-09-02 12:47:03.069 [INF] CRTR: Processed channels=0 updates=272 nodes=0 in last 59.999710346s
2020-09-02 12:47:07.338 [INF] DISC: Broadcasting 388 new announcements in 18 sub batches
2020-09-02 12:48:03.069 [INF] CRTR: Processed channels=0 updates=170 nodes=3 in last 59.999843239s
2020-09-02 12:48:37.433 [INF] DISC: Broadcasting 219 new announcements in 17 sub batches
2020-09-02 12:48:54.682 [INF] CRTR: Pruning channel graph using block 0000000000000000000e65c5c2126e7c902fa09351b8087ab331182eeabebfab (height=646431)
2020-09-02 12:48:56.056 [INF] NTFN: New block: height=646431, sha=0000000000000000000e65c5c2126e7c902fa09351b8087ab331182eeabebfab
2020-09-02 12:48:56.059 [ERR] LTND: Unable to lookup witness: no witnesses
2020-09-02 12:48:56.060 [INF] UTXN: Attempting to graduate height=646431: num_kids=0, num_babies=0
2020-09-02 12:48:56.413 [INF] HSWC: Garbage collected 1 shared secret hashes at height=646431
2020-09-02 12:48:56.467 [INF] CNCT: ChannelArbitrator(25194f104e65a27cfd6867ac8e8708c9cb898e1212a51ff1862ffb03cc1e340b:1): force closing chan
2020-09-02 12:48:56.611 [INF] HSWC: Removing channel link with ChannelID(0b341ecc03fb2f86f11fa512128e89cbc908878eac6768fd7ca2654e104f1924)
2020-09-02 12:48:57.822 [INF] CRTR: Block 0000000000000000000e65c5c2126e7c902fa09351b8087ab331182eeabebfab (height=646431) closed 1 channels
2020-09-02 12:49:00.967 [INF] CNCT: Broadcasting force close transaction d9ad30a963638a6efe405b608e1746b55092cb877e620d7135fe083becadbc51, ChannelPoint(25194f104e65a27cfd6867ac8e8708c9cb898e1212a51ff1862ffb03cc1e340b:1): (*wire.MsgTx)(0x40016b6200)({
 Version: (int32) 2,
 TxIn: ([]*wire.TxIn) (len=1 cap=1) {
  (*wire.TxIn)(0x40072170e0)({
   PreviousOutPoint: (wire.OutPoint) 25194f104e65a27cfd6867ac8e8708c9cb898e1212a51ff1862ffb03cc1e340b:1,
   SignatureScript: ([]uint8) <nil>,
   Witness: (wire.TxWitness) (len=4 cap=4) {
    ([]uint8) <nil>,
    ([]uint8) (len=72 cap=144) {
     00000000  30 45 02 21 00 cb 77 1a  b5 0a 77 19 78 f8 6e d1  |0E.!..w...w.x.n.|
     00000010  5b d3 0d f2 c5 09 b8 0a  f1 cd 7b e4 d5 37 54 ca  |[.........{..7T.|
     00000020  03 8c fd 45 41 02 20 00  98 e9 33 4e 9b 54 72 1c  |...EA. ...3N.Tr.|
     00000030  03 46 f9 80 7b f3 3f 2c  ff 9c 9b bf 9d bf fb d6  |.F..{.?,........|
     00000040  57 0f d7 89 b8 b5 9f 01                           |W.......|
    },
    ([]uint8) (len=72 cap=144) {
     00000000  30 45 02 21 00 c7 1b 89  a9 bb 26 e3 70 62 da 92  |0E.!......&.pb..|
     00000010  18 79 e4 b3 c6 2d cd d2  23 55 33 82 6b 6d 69 e5  |.y...-..#U3.kmi.|
     00000020  4f e5 15 e8 ed 02 20 1a  0f c3 a1 e3 78 c9 60 02  |O..... .....x.`.|
     00000030  23 33 20 47 70 0b 4c a2  71 0a 75 5c 8c f0 ef 90  |#3 Gp.L.q.u\....|
     00000040  5a b2 d8 b8 21 92 e2 01                           |Z...!...|
    },
    ([]uint8) (len=71 cap=500) {
     00000000  52 21 02 10 09 28 9d 07  68 3c dc 41 e5 b6 c5 45  |R!...(..h<.A...E|
     00000010  2d e9 ae e4 2f 24 5a bd  7f c5 ca 2b 7f 81 87 f7  |-.../$Z....+....|
     00000020  1f 5f f9 21 03 fd 59 d9  e6 f3 cc 0b 54 bb 88 11  |._.!..Y.....T...|
     00000030  8b 8e b1 aa 2d 5e f7 f3  b2 35 7f 05 9d 84 d4 0e  |....-^...5......|
     00000040  ab 89 94 d2 d9 52 ae                              |.....R.|
    }
   },
   Sequence: (uint32) 2160573938
  })
 },
 TxOut: ([]*wire.TxOut) (len=2 cap=2) {
  (*wire.TxOut)(0x4004b83480)({
   Value: (int64) 18918,
   PkScript: ([]uint8) (len=22 cap=22) {
    00000000  00 14 34 61 63 e4 5b 89  ae dd 6b 9d 6f 3d bf 58  |..4ac.[...k.o=.X|
    00000010  17 fd c8 2e d4 61                                 |.....a|
   }
  }),
  (*wire.TxOut)(0x4004b834a0)({
   Value: (int64) 151877,
   PkScript: ([]uint8) (len=34 cap=34) {
    00000000  00 20 64 f9 21 c5 22 fc  e3 38 b4 97 59 36 eb a5  |. d.!."..8..Y6..|
    00000010  6d c4 1f 4a 2d 1d 25 d9  18 9b 7d 9b 3b d8 cc cf  |m..J-.%...}.;...|
    00000020  4a db                                             |J.|
   }
  })
 },
 LockTime: (uint32) 540252184
})

2020-09-02 12:49:01.181 [INF] LNWL: Inserting unconfirmed transaction d9ad30a963638a6efe405b608e1746b55092cb877e620d7135fe083becadbc51

```

### Your environment
```
lnd version 0.11.0-beta commit=v0.11.0-beta
Bitcoin Core Daemon version v0.18.0
Linux rock64 4.4.132-1075-rockchip-ayufan-ga83beded8524 #1 SMP Thu Jul 26 08:22:22 UTC 2018 aarch64 GNU/Linux
```
Full log available here (password protected, send me a slack message and I'll send you the password)

[lnd.log.80.zip](https://github.com/lightningnetwork/lnd/files/5214538/lnd.log.80.zip)

### PR DESCRIPTION
**Pull request by Crypt-iQ**
With this pull request, we store a boolean per channel under the new `lastWasRevokeKey`. This key determines if the last material update our channel *sent* was a `CommitSig` or `RevokeAndAck`.  When the link calls `SignNextCommitment`, `false` is stored under the key.  When the link calls `RevokeAndAck`, `true` is stored under the key.

For new channels with no updates yet exchanged, nothing is stored and so the key contains `nil`.

For existing channels updating to this patch, the default `nil` is also stored.  The next update sent will store something under the key. 

Fixes https://github.com/lightningnetwork/lnd/issues/4799
Fixes #4613

### CODE REVIEW

File: htlcswitch/link_test.go [hunk 1]
```diff
@@ -180,6 +180,290 @@ func createInterceptorFunc(prefix, receiver string, messages []expectedMessage,
...
+//   -----rev----x
+//   -----sig----x
+func TestChannelLinkRevThenSig(t *testing.T) {
+	t.Parallel()
+
+	const chanAmt = btcutil.SatoshiPerBitcoin * 5
+	const chanReserve = btcutil.SatoshiPerBitcoin * 1
+	aliceLink, bobChannel, batchTicker, start, cleanUp, restore, err :=
+		newSingleLinkTestHarness(chanAmt, chanReserve)
+	if err != nil {
```
Comment (cfromknecht) [inline htlcswitch/link_test.go:202]: a few places in these tests where we can start to use `require`

File: lnwallet/channel.go [hunk 2]
```diff
@@ -3892,16 +3896,41 @@ func (lc *LightningChannel) ProcessChanSyncMsg(
...
+			}
+		}
+
+		var commitUpdates []lnwire.Message
+
 		// Next, we'll need to send over any updates we sent as part of
 		// this new proposed commitment state.
 		for _, logUpdate := range commitDiff.LogUpdates {
-			updates = append(updates, logUpdate.UpdateMsg)
+			commitUpdates = append(commitUpdates, logUpdate.UpdateMsg)
```
Comment (cfromknecht) [inline lnwallet/channel.go:3921]: is the `RevokeAndAck` a part of `commitDiff.LogUpdates`?

File: lnwallet/channel.go [hunk 3]
```diff
@@ -3892,16 +3896,41 @@ func (lc *LightningChannel) ProcessChanSyncMsg(
...
 		}
 
 		// With the batch of updates accumulated, we'll now re-send the
 		// original CommitSig message required to re-sync their remote
 		// commitment chain with our local version of their chain.
-		updates = append(updates, commitDiff.CommitSig)
+		commitUpdates = append(commitUpdates, commitDiff.CommitSig)
+
+		if lastWasRevoke {
```
Comment (cfromknecht) [inline lnwallet/channel.go:3929]: a comment here would be useful to explain how these two orderings differ

**Review by cfromknecht — COMMENTED** [review]
first pass looks solid! nice to see that the fix was straightforward. a few nits but otherwise looks good

File: lnwallet/channel.go (hunk 2 above)
Comment (Crypt-iQ) [reply lnwallet/channel.go:3921]: nope it's already added in `updates`

File: htlcswitch/link_test.go [hunk 4]
```diff
@@ -180,6 +180,290 @@ func createInterceptorFunc(prefix, receiver string, messages []expectedMessage,
...
+	comSig, ok := msg.(*lnwire.CommitSig)
+	if !ok {
+		t.Fatalf("expected CommitSig, got %T", msg)
+	}
+
+	if len(comSig.HtlcSigs) != 2 {
+		t.Fatalf("expected 2 htlc sigs, got %d", len(comSig.HtlcSigs))
+	}
+
+	cleanUp = alice.restart(false, true)
```
Comment (halseth) [inline htlcswitch/link_test.go:279]: add comment

File: htlcswitch/link_test.go [hunk 5]
```diff
@@ -180,6 +180,290 @@ func createInterceptorFunc(prefix, receiver string, messages []expectedMessage,
...
+	if err != nil {
+		t.Fatalf("unable to retrieve chan sync msg: %v", err)
+	}
+	ctx.aliceLink.HandleChannelUpdate(bobReest)
+
+	// ------rev---->
+	ctx.receiveRevAndAckAliceToBob()
+
+	// ------add---->
+	// We don't send this to Bob as his channel state already contains this,
```
Comment (halseth) [inline htlcswitch/link_test.go:308]: why is this? Since Bob was not restarted?

Shouldn't bob handle receiving this from Alice?

File: htlcswitch/link_test.go [hunk 6]
```diff
@@ -180,6 +180,290 @@ func createInterceptorFunc(prefix, receiver string, messages []expectedMessage,
...
+	case <-time.After(15 * time.Second):
+		t.Fatalf("did not receive message")
+	}
+
+	_, ok = msg.(*lnwire.UpdateAddHTLC)
+	if !ok {
+		t.Fatalf("expected UpdateAddHTLC, got %T", msg)
+	}
+
+	// ------sig---->
```
Comment (halseth) [inline htlcswitch/link_test.go:460]: seems like the two tests could be combined into one, with subtests that just switches the order at the end here?

**Review by halseth — CHANGES_REQUESTED** [review]

File: htlcswitch/link_test.go (hunk 5 above)
Comment (Crypt-iQ) [reply htlcswitch/link_test.go:308]: We'll get a duplicate htlc id if Bob accepts (0) while Bob is expecting 1. It's because Bob is represented as `LightningChannel`

File: htlcswitch/link_test.go (hunk 6 above)
Comment (Crypt-iQ) [reply htlcswitch/link_test.go:460]: `TestChannelLinkRevThenSig` tests:
```
//   <----sig-----
//   -----rev----x
//   -----sig----x
```

`TestChannelLinkSigThenRev` tests:
```
//   -----sig----x
//   <----sig-----
//   -----rev----x
```



The order is slightly different in the 2nd case. We can't test:
```
//   <----sig-----
//   -----sig----x
//   -----rev----x
```
because we can't control when `RevokeAndAck` is sent in the link, but we can for `CommitSig` with the force ticker.  So when Bob sends a signature, Alice will always reply with a revocation first, there's no way to hold off on that. So to get the desired ordering in the second case, Alice sends her sig (which Bob doesn't receive), Bob sends his sig, and then Alice replies with a revoke (which Bob doesn't receive). Since they're different in this way, I don't think they should go in subtests.

File: channeldb/channel.go [hunk 7]
```diff
@@ -1526,6 +1535,15 @@ func (c *OpenChannel) UpdateCommitment(newCommitment *ChannelCommitment,
 				"updates: %v", err)
 		}
 
+		var b2 bytes.Buffer
+		if err := WriteElements(&b2, true); err != nil {
+			return err
+		}
+
+		if err := chanBucket.Put(lastWasRevokeKey, b2.Bytes()); err != nil {
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply channeldb/channel.go:1545]: Comment added

File: htlcswitch/link_test.go (hunk 4 above)
Comment (Crypt-iQ) [reply htlcswitch/link_test.go:273]: Comment added

File: htlcswitch/link_test.go (hunk 1 above)
Comment (Crypt-iQ) [reply htlcswitch/link_test.go]: Using require

File: lnwallet/channel.go [hunk 8]
```diff
@@ -3892,16 +3892,30 @@ func (lc *LightningChannel) ProcessChanSyncMsg(
...
 		// commitment chain with our local version of their chain.
-		updates = append(updates, commitDiff.CommitSig)
+		commitUpdates = append(commitUpdates, commitDiff.CommitSig)
+
+		// NOTE: If a revocation is not owed, then updates is empty.
+		if lc.channelState.LastWasRevoke {
+			// If lastWasRevoke is set to true, a revocation was last and we
+			// need to reorder the updates so that the revocation stored in
+			// updates comes after the LogUpdates+CommitSig.
+			updates = append(commitUpdates, updates...)
```
Comment (halseth) [inline lnwallet/channel.go:3913]: I find this a bit hard to follow, possibly because we are assuming something about the updates that have been added in an earlier switch statement.

Could an alternative be to split this up
```go
var owedRevocation, owedCommitUpdates []lnwire.Message

switch {
case ...:
    owedRevocation = append(owedRevocation, revokeAndAck)
}

switch {
case ...:
    owedCommitUpdates = append(owedCommitUpdates, commitSig)
}

var updates []lnwire.Message
if lastWasRevoke {
    updates = append(owedCommitUpdates, owedRevocation)
else {
    updates = append(owedRevocation, owedCommitUpdates) 
}

```

File: lnwallet/channel.go [hunk 9]
```diff
@@ -3892,16 +3892,30 @@ func (lc *LightningChannel) ProcessChanSyncMsg(
 			return nil, nil, nil, err
 		}
 
+		var commitUpdates []lnwire.Message
```
Comment (halseth) [inline lnwallet/channel.go:3895]: I wonder if we could also encode the `OweCommitment` case above using this new key?

File: htlcswitch/link_test.go (hunk 5 above)
Comment (halseth) [reply htlcswitch/link_test.go:298]: I see. Would i make sense to also restart Bob?

**Review by halseth — CHANGES_REQUESTED** [review]

File: lnwallet/channel.go (hunk 9 above)
Comment (Crypt-iQ) [reply lnwallet/channel.go:3895]: If `OweCommitment` is called, then this block of code cannot be called

Comment (Crypt-iQ) [reply lnwallet/channel.go:3895]: Sorry, I don't understand the suggestion

File: lnwallet/channel.go (hunk 8 above)
Comment (Crypt-iQ) [reply lnwallet/channel.go:3913]: I chose this way because this does not account for `OweCommitment`.  Would it then be added to `owedRevocation`? - Then the name doesn't describe what it does.  The switch `case msg.NextLocalCommitHeight == remoteTipHeight` cannot be called if `OweCommitment` is called so that doesn't change our current logic if a revocation exists or not.

File: lnwallet/channel.go (hunk 3 above)
Comment (Crypt-iQ) [reply lnwallet/channel.go]: Comment added

File: htlcswitch/link_test.go (hunk 5 above)
Comment (Crypt-iQ) [reply htlcswitch/link_test.go]: Yes, restarting Bob now. Makes the tests cleaner

**Review by cfromknecht — APPROVED** [review]
LGTM, nice find on this being absent from the spec and a simple fix 🥂

File: lnwallet/channel.go (hunk 9 above)
Comment (Roasbeef) [reply lnwallet/channel.go:3895]: I think he means modify the logic above that determines if we need to send a new sig to use the new value added in the change. IMO we should just try to make the smallest change here, so we can isolate the impact of the diff itself.

File: lnwallet/channel.go (hunk 8 above)
Comment (Roasbeef) [reply lnwallet/channel.go:3917]: > I chose this way because this does not account for OweCommitment

So won't this end up _re ordering_ the `OweCommitment` sig if we need to generate+send it?

Comment (Roasbeef) [reply lnwallet/channel.go:3917]: > The switch case msg.NextLocalCommitHeight == remoteTipHeight cannot be called if OweCommitment is called so that doesn't change our current logic if a revocation exists or not.

Ah ok, make sense.

**Review by Roasbeef — APPROVED** [review]
Excellent find and fix!

To our knowledge, there's just one lingering issue in this area. However, we know this PR in itself improves the situation, so it makes sense to move forward with it.
